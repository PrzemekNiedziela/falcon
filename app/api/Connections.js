// @flow
import fs from 'fs';
import Store from 'electron-store';
import _ from 'lodash';
import { verifySqlite } from './Database';
import type { LoginSavedDatabaseType } from '../types/LoginSavedDatabaseType';

/**
 * Handles saving/loading of connections to electron-store.
 * Verifies nicknames, paths, and validity of file
 */
export default class Connections {
  store = new Store();

  /**
   * @return  saved databases local to the machine. If no such array exists,
   * returns an empty array
   */
  getSavedDatabases = (): Array<LoginSavedDatabaseType> =>
    this.store.get('savedDatabases') || [];

  /**
   * Saves the database to local storage. Throws an error if nickname is empty
   * or if no database is found
   * @return an array containing the new database
   */
  saveDatabase = async (
    databaseNickname: string,
    databasePath: string
  ): Promise<Array<LoginSavedDatabaseType>> => {
    const newDatabase = {
      nickname: databaseNickname,
      path: databasePath
    };
    const savedDatabases = this.store.get('savedDatabases') || [];
    if (await this.validateSave(newDatabase)) {
      savedDatabases.push(newDatabase);
      this.store.set('savedDatabases', savedDatabases);
      return savedDatabases;
    }
    throw new Error('Saved databases need a nickname and a valid database');
  };

  /**
   * Validates a potentialDatabase. Database file must exist,
   * @return true if potentialDatabase valid, false if not
   */
  validateSave = async (
    potentialDatabase: LoginSavedDatabaseType
  ): Promise<boolean> => {
    const savedDatabases = this.store.get('savedDatabases') || [];
    const { nickname, path } = potentialDatabase;
    const databaseFileExists = await Connections.validateDatabaseFilePath(path);
    return (
      databaseFileExists &&
      nickname !== '' &&
      Connections.isDatabaseSaved(savedDatabases, potentialDatabase)
    );
  };

  /**
   * Deletes a savedDatabase from local storage
   * @return an array that does not contain that database
   */
  deleteSavedDatabase = (savedDatabase: LoginSavedDatabaseType): Array<LoginSavedDatabaseType> => {
    const savedDatabases = _.cloneDeep(this.store.get('savedDatabases')).filter(e => _.isEqual(e, savedDatabase));
    this.store.set('savedDatabases', savedDatabases);
    return savedDatabases;
  };

  /**
   * Validates a database file's path
   * @return true if databasefile exists and is .db, .sqlite, or sqlite3
   */
  static validateDatabaseFilePath(databasePath: string): Promise<boolean> {
    const fileExtension = databasePath.substring(databasePath.lastIndexOf('.'));
    return new Promise((resolve, reject) => {
      fs.stat(databasePath, (err, stats) => {
        if (typeof stats === 'undefined') {
          resolve(false);
          return;
        }
        if (err) {
          if (err.code === 'ENOENT') {
            resolve(false);
            return;
          }
          reject(err);
          return;
        }
        if (
          stats.isFile() &&
          (fileExtension === '.db' ||
            fileExtension === '.sqlite' ||
            fileExtension === '.sqlite3')
        ) {
          resolve(true);
        }
      });
    });
  }

  /**
   * Checks if a given database file has a problem
   * @return true if no problem exists, the error if there is a problem
   */
  static async validateConnection(databasePath: string): Promise<string | true> {
    return verifySqlite(databasePath);
  }

  /**
   * Checks if a database is saved in savedDatabases
   * @return true if saved, false if not
   */
  static isDatabaseSaved(
    savedDatabases: Array<LoginSavedDatabaseType>,
    database: LoginSavedDatabaseType
  ): boolean {
    return savedDatabases.some(e => _.isEqual(e, database));
  }
}
