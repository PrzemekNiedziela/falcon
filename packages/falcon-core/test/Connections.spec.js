// @flow
/* eslint no-await-in-loop: 0 */
import Connections from '../src/config/ConnectionManager';

async function connectionFactory(connections, connectionCount: number = 1) {
  const array = new Array(connectionCount);
  for (let i = 0; i < array.length; i++) {
    array[i] = await connections.add({
      id: `test-id-${i + 1}`,
      type: 'sqlite',
      name: `test-connection-${i + 1}`,
      database: '/Users/amila/Desktop/demo.sqlite'
    });
  }
  return Promise.all(array);
}

describe('Connections', function testConnections() {
  beforeEach(async () => {
    this.connections = new Connections();
    await connectionFactory(this.connections, 3);
  });

  afterEach(async () => {
    await this.connections.removeAll();
  });

  it('should get all connections', async () => {
    const connections = await this.connections.getAll();
    expect(connections).toMatchSnapshot();
  });

  it('should delete a single connection', async () => {
    const connections = await this.connections.getAll();
    const connectionIdToDelete = connections[0].id;
    await this.connections.remove(connectionIdToDelete);
    const newConnections = await this.connections.getAll();
    expect(newConnections).toMatchSnapshot();
  });

  it('should update a single connection', async () => {
    const connections = await this.connections.getAll();
    const connectionIdToDelete = connections[0].id;
    await this.connections.update(connectionIdToDelete, {
      id: 'test-id-foo',
      name: 'test-connection-foo',
      password: 'test-password',
      type: 'sqlite'
    });
    const newConnections = await this.connections.getAll();
    expect(newConnections).toMatchSnapshot();
  });

  it('should get a single connection', async () => {
    const connections = await this.connections.getAll();
    const connectionId = connections[2].id;
    const connection = await this.connections.get(connectionId);
    expect(connection).toMatchSnapshot();
  });

  it('should perform basic validation', async () => {
    expect(() => {
      this.connections.validateBeforeCreation({
        id: 12,
        database: 'aJ@#LJ#@KL$KL@#sdf',
        type: 'sqlite'
      });
    }).toThrowErrorMatchingSnapshot();
    expect(() => {
      this.connections.validateBeforeCreation({
        id: 12,
        database: '/usr/foo',
        type: 'sqlite'
      });
    }).toThrowErrorMatchingSnapshot();
    expect(() => {
      this.connections.validateBeforeCreation({
        id: 'foo',
        database: '/usr/foo',
        type: 'sqlite'
      });
    }).toThrowErrorMatchingSnapshot();
    expect(() => {
      this.connections.validateBeforeCreation({
        id: 'foo',
        name: 'foo',
        database: '/usr/foo',
        type: 'sqlite'
      });
    }).toThrowErrorMatchingSnapshot();
    expect(() => {
      this.connections.validateBeforeCreation({
        id: 'foo',
        name: 'foo',
        database: '/usr/local/bin/npm',
        type: 'sqlite'
      });
    }).toThrowErrorMatchingSnapshot();
  });

  it('should check if a sqlite file is valid or not', async () => {
    expect(
      this.connections.validateBeforeCreation({
        id: 'foo',
        name: 'foo',
        database: '/Users/amila/Desktop/demo.sqlite',
        type: 'sqlite'
      })
    ).toEqual(undefined);
  });
});
