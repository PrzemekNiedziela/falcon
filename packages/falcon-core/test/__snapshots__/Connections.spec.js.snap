// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connections should delete a single connection 1`] = `
Array [
  Object {
    "color": "gray",
    "database": "/Users/amila/Desktop/demo.sqlite",
    "id": "test-id-2",
    "name": "test-connection-2",
    "type": "sqlite",
  },
  Object {
    "color": "gray",
    "database": "/Users/amila/Desktop/demo.sqlite",
    "id": "test-id-3",
    "name": "test-connection-3",
    "type": "sqlite",
  },
]
`;

exports[`Connections should get a single connection 1`] = `
Object {
  "color": "gray",
  "database": "/Users/amila/Desktop/demo.sqlite",
  "id": "test-id-3",
  "name": "test-connection-3",
  "type": "sqlite",
}
`;

exports[`Connections should get all connections 1`] = `
Array [
  Object {
    "color": "gray",
    "database": "/Users/amila/Desktop/demo.sqlite",
    "id": "test-id-1",
    "name": "test-connection-1",
    "type": "sqlite",
  },
  Object {
    "color": "gray",
    "database": "/Users/amila/Desktop/demo.sqlite",
    "id": "test-id-2",
    "name": "test-connection-2",
    "type": "sqlite",
  },
  Object {
    "color": "gray",
    "database": "/Users/amila/Desktop/demo.sqlite",
    "id": "test-id-3",
    "name": "test-connection-3",
    "type": "sqlite",
  },
]
`;

exports[`Connections should perform basic validation 1`] = `"Failed validation: [{\\"message\\":\\"\\\\\\"id\\\\\\" must be a string\\",\\"fieldName\\":\\"id\\"},{\\"message\\":\\"\\\\\\"name\\\\\\" is required\\",\\"fieldName\\":\\"name\\"},{\\"message\\":\\"\\\\\\"database\\\\\\" needs to be a file\\",\\"fieldName\\":\\"database\\"},{\\"message\\":\\"\\\\\\"database\\\\\\" does not exist\\",\\"fieldName\\":\\"database\\"},{\\"message\\":\\"\\\\\\"database\\\\\\" is not valid\\",\\"fieldName\\":\\"database\\"}]"`;

exports[`Connections should perform basic validation 2`] = `"Failed validation: [{\\"message\\":\\"\\\\\\"id\\\\\\" must be a string\\",\\"fieldName\\":\\"id\\"},{\\"message\\":\\"\\\\\\"name\\\\\\" is required\\",\\"fieldName\\":\\"name\\"},{\\"message\\":\\"\\\\\\"database\\\\\\" does not exist\\",\\"fieldName\\":\\"database\\"},{\\"message\\":\\"\\\\\\"database\\\\\\" is not valid\\",\\"fieldName\\":\\"database\\"}]"`;

exports[`Connections should perform basic validation 3`] = `"Failed validation: [{\\"message\\":\\"\\\\\\"name\\\\\\" is required\\",\\"fieldName\\":\\"name\\"},{\\"message\\":\\"\\\\\\"database\\\\\\" does not exist\\",\\"fieldName\\":\\"database\\"},{\\"message\\":\\"\\\\\\"database\\\\\\" is not valid\\",\\"fieldName\\":\\"database\\"}]"`;

exports[`Connections should perform basic validation 4`] = `"Failed validation: [{\\"message\\":\\"\\\\\\"database\\\\\\" does not exist\\",\\"fieldName\\":\\"database\\"},{\\"message\\":\\"\\\\\\"database\\\\\\" is not valid\\",\\"fieldName\\":\\"database\\"}]"`;

exports[`Connections should perform basic validation 5`] = `"Failed validation: [{\\"message\\":\\"\\\\\\"database\\\\\\" is not valid\\",\\"fieldName\\":\\"database\\"}]"`;

exports[`Connections should update a single connection 1`] = `
Array [
  Object {
    "color": "gray",
    "database": "/Users/amila/Desktop/demo.sqlite",
    "id": "test-id-1",
    "name": "test-connection-1",
    "type": "sqlite",
  },
  Object {
    "color": "gray",
    "database": "/Users/amila/Desktop/demo.sqlite",
    "id": "test-id-2",
    "name": "test-connection-2",
    "type": "sqlite",
  },
  Object {
    "color": "gray",
    "database": "/Users/amila/Desktop/demo.sqlite",
    "id": "test-id-3",
    "name": "test-connection-3",
    "type": "sqlite",
  },
]
`;
